tutorials/intermediate/blitting.py

```diff
@@ -0,0 +1,210 @@
+"""
+=================
+Blitting tutorial
+=================
+
+'Blitting' is a `standard technique
+<https://en.wikipedia.org/wiki/Bit_blit>`__ in raster graphics that,
+in the context of Matplotlib, can be used to (drastically) improve
+performance of interactive figures. For example, the
+:mod:`~.animation` and :mod:`~.widgets` modules use blitting
+internally. Here, we demonstrate how to implement your own blitting, outside
+of these classes.
+
+The source of the performance gains is simply not re-doing work we do
+not have to.  If the limits of an Axes have not changed, then there is
+no need to re-draw all of the ticks and tick-labels (particularly
+because text is one of the more expensive things to render).
+
+The procedure to save our work is roughly:
+
+- draw the figure, but exclude any artists marked as 'animated'
+- save a copy of the RBGA buffer
+
+In the future, to update the 'animated' artists we
+
+- restore our copy of the RGBA buffer
+- redraw only the animated artists
+- show the resulting image on the screen
+
+thus saving us from having to re-draw everything which is _not_
+animated.  One consequence of this procedure is that your animated
+artists are always drawn at a higher z-order than the static artists.
+
+Not all backends support blitting.  You can check if a given canvas does via
+the `.FigureCanvasBase.supports_blit` property.
+
+Minimal example
+---------------
+
+We can use the `.FigureCanvasAgg` methods
+`~.FigureCanvasAgg.copy_from_bbox` and
+`~.FigureCanvasAgg.restore_region` in conjunction with setting
+``animated=True`` on our artist to implement a minimal example that
+uses blitting to accelerate rendering
+
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+x = np.linspace(0, 2 * np.pi, 100)
+
+fig, ax = plt.subplots()
+# animated=True tells matplotlib to only draw the artist when we
+# explicitly request it
+(ln,) = ax.plot(x, np.sin(x), animated=True)
+
+# make sure the window is raised, but the script keeps going
+plt.show(block=False)
+# stop to admire our empty window axes and ensure it is rendered at
+# least once
+plt.pause(0.1)
```

Comment:
This sequence now makes a lot more sense for a beginner. 

I think this is good to go in as-is.

However, there's still some easily confusing things here (like, you can remove the `pause` if you run this code line-by-line and it will work, but if you paste all the lines together into ipython at once it fails, presumably due to not being in interactive mode). Again, I don't think this should block this tutorial from going in (something about perfect and enemies and good), as it's a valuable addition. But it took me an hour just now to figure out how to make any changes to this example without breaking it (and I had to read the source and inspect  private runtime attributes of `pyplot` to do so). So maybe consider the following version (if I'm not still totally off-base about how to use the interactive interface)?

```python
# "draw" all non-interactive artists onto the canvas
fig.canvas.draw()
# make sure our backend knows to display our canvas onto the screen
plt.show(block=False)
# let the event loop spin quickly, to process e.g. window resizing events
# (this is mandatory for tiling window managers)
fig.canvas.start_event_loop(0.1)  # runs GUI for 0.1s
```

Saying "fig.canvas.start_event_loop" instead of "plt.pause" adds even more semantic clarity on top of your explicit call to "plt.show" (which had already made this example much more clear, IMO). And it seems like the only other thing in `pyplot.pause` is a call to `fig.canvas.draw_idle`, which is necessary to cache the renderer so that later calls to `draw_artist` will work.

Reply:
Which backend were you using that didn't work when pasted in to IPython and what OS?  I get the expected behavior with Qt5Agg.

What were you trying to change that went badly?

I would rather not drop to the lower-level things at this point if we can avoid it.

But this does suggest I should add `start_event_loop` to https://github.com/matplotlib/matplotlib/pull/4779 :facepalm:

Reply:
Ah, I did https://github.com/matplotlib/matplotlib/pull/4779/files#diff-c048f9588ed99e2733a770d35b1b8b39R105 the diff was just hidden by default.

Reply:
Yes, that mention in the new docs is the only reason I was ever able to get this to work. The sequence of events of "things that went wrong" is complicated enough that I pasted it below so it doesn't get buried in comments here.